apps/emqx_durable_storage/src/emqx_ds_storage_layer.erl

```diff
@@ -444,15 +449,12 @@ update_iterator(
 next(Shard, Iter = #{?tag := ?IT, ?generation := GenId, ?enc := GenIter0}, BatchSize, Now) ->
     case generation_get(Shard, GenId) of
         #{module := Mod, data := GenData} ->
-            Current = generation_current(Shard),
-            case Mod:next(Shard, GenData, GenIter0, BatchSize, Now) of
-                {ok, _GenIter, []} when GenId < Current ->
-                    %% This is a past generation. Storage layer won't write
-                    %% any more messages here. The iterator reached the end:
-                    %% the stream has been fully replayed.
-                    {ok, end_of_stream};
+            IsCurrent = GenId =:= generation_current(Shard),
```

Comment:
I wonder if this loss of write idempotency could be alleviated by picking the "current" generation according to `Now` here, instead of the most recent storage state? I have a feeling this also relaxes the requirement to read from leaders, but am not 100% sure.

The risk here I guess is having `Now` pointing to a dropped generation (from the _storage state_ perspective), but picking the earliest alive in that case should work the same I guess.